pkg/controller/podautoscaler/replica_calculator.go

```diff
@@ -365,6 +365,12 @@ func groupPods(pods []*v1.Pod, metrics metricsclient.PodMetricsInfo, resource v1
 		if pod.DeletionTimestamp != nil || pod.Status.Phase == v1.PodFailed {
 			continue
 		}
+		// Unschedulable pods are always ignored.
+		_, scheduled := podutil.GetPodCondition(&pod.Status, v1.PodScheduled)
```

Comment:
What if there is no scheduled condition? Isn't pod unscheduled in that case?
Shouldn't we also ignore pods that were scheduled but are not running yet?
Maybe we should use pod.Status.Phase instead?

Reply:
> Maybe we should use pod.Status.Phase instead?

Yes, I think you're right.

It's very important that, on scale up, pods that are not reporting metrics or are not ready are considered as using 0%.  This ensures that the HPA doesn't runaway because pods are slow to start or the metrics pipeline is slow (it is).  We already have this case which will include ignored pods as 0% in the metrics: https://github.com/kubernetes/kubernetes/blob/efdc3fb0a2938b72e4de02f147ab3329895ff018/pkg/controller/podautoscaler/replica_calculator.go#L118-L123

Since we already consider all ignored pods as 0% in addition to the missing metric pods, I think there is no harm in including **all** pending pods in the ignored set.

Reply:
Done.

Reply:
I meant that I'm not sure if pod after creation has condition PodScheduled. Even if it has the condition then scheduled.Reason won't be equal to v1.PodReasonUnschedulable. If there is a lot of pods to be scheduled then it can take some time before condition is added by scheduler. Maybe it would be better to use pod.Spec.NodeName != nil instead. All scheduled pods have this field set.